drivers/nrf_wifi/fw_if/umac_if/src/fmac_promisc.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * @brief File containing promiscuous mode
+ * support functions for the FMAC IF Layer
+ * of the Wi-Fi driver.
+ */
+
+#include "osal_api.h"
+#include "fmac_structs.h"
+#include "fmac_promisc.h"
+
+bool nrf_wifi_util_check_filt_setting(struct nrf_wifi_fmac_vif_ctx *vif,
+				      unsigned short *frame_control)
+{
+	bool filter_check = false;
+	struct nrf_wifi_fmac_frame_ctrl *frm_ctrl =
+	(struct nrf_wifi_fmac_frame_ctrl *)frame_control;
+
+	/**
+	 * The different filter settings for 802.11 packets within the driver
+	 * is a bit map. The description of the different valid values for
+	 * the case statements are as follows:
+	 * 0x2 - Enable management packets only
+	 * 0x4 - Enable data packets only
+	 * 0x8 - Enable control packets only
+	 * 0x6 - Enable data and management packets
+	 * 0xa - Enable control and management packets
+	 * 0xc - Enable data and control packets
+	 * The bit map setting is checked against the packet type field and
+	 * the relevant filtered packet is sent up the stack.  if bit 0 is set
+	 * in the driver bitmap, all packet types are allowed to be sent upstack.
+	 **/
+	switch (vif->packet_filter) {
+		case 0x2:
+			if (frm_ctrl->type == NRF_WIFI_MGMT_PKT_TYPE) {
+				filter_check = true;
+			}
+			break;
+		case 0x4:
+			if (frm_ctrl->type == NRF_WIFI_DATA_PKT_TYPE) {
+				filter_check = true;
+			}
+			break;
+		case 0x6:
+			if ((frm_ctrl->type == NRF_WIFI_DATA_PKT_TYPE) ||
+			    (frm_ctrl->type == NRF_WIFI_MGMT_PKT_TYPE)) {
+				filter_check = true;
+			}
+			break;
+		case 0x8:
+			if (frm_ctrl->type == NRF_WIFI_CTRL_PKT_TYPE) {
+				filter_check = true;
+			}
+			break;
+		case 0xa:
+			if ((frm_ctrl->type == NRF_WIFI_CTRL_PKT_TYPE) ||
+			    (frm_ctrl->type == NRF_WIFI_MGMT_PKT_TYPE)) {
+				filter_check = true;
+			}
+			break;
+		case 0xc:
+			if ((frm_ctrl->type == NRF_WIFI_DATA_PKT_TYPE) ||
+			    (frm_ctrl->type == NRF_WIFI_CTRL_PKT_TYPE)) {
+				filter_check = true;
+			}
+			break;
+		/* all other packet_filter cases - bit 0 is set and hence all
+		 * packet types are allowed or in the case of 0xE - all packet
+		 * type bits are enabled */
+		default:
+			filter_check = true;
+			break;
+	}
+	return filter_check;
+}
```

Comment:
This is what my co-pilot gave me :), it's definitely more readable.

```
#include <stdbool.h>

enum nrf_wifi_pkt_type {
	NRF_WIFI_MGMT_PKT_TYPE,
	NRF_WIFI_DATA_PKT_TYPE,
	NRF_WIFI_CTRL_PKT_TYPE
};

enum nrf_wifi_packet_filter {
	FILTER_MGMT_ONLY = 0x2,
	FILTER_DATA_ONLY = 0x4,
	FILTER_DATA_MGMT = 0x6,
	FILTER_CTRL_ONLY = 0x8,
	FILTER_CTRL_MGMT = 0xa,
	FILTER_DATA_CTRL = 0xc
};

struct nrf_wifi_fmac_frame_ctrl {
	enum nrf_wifi_pkt_type type;
};

struct nrf_wifi_fmac_vif_ctx {
	unsigned short packet_filter;
};

bool nrf_wifi_util_check_filt_setting(struct nrf_wifi_fmac_vif_ctx *vif,
									  unsigned short *frame_control)
{
	bool filter_check = false;
	struct nrf_wifi_fmac_frame_ctrl *frm_ctrl =
		(struct nrf_wifi_fmac_frame_ctrl *)frame_control;

	switch (vif->packet_filter) {
		case FILTER_MGMT_ONLY:
			filter_check = (frm_ctrl->type == NRF_WIFI_MGMT_PKT_TYPE);
			break;
		case FILTER_DATA_ONLY:
			filter_check = (frm_ctrl->type == NRF_WIFI_DATA_PKT_TYPE);
			break;
		case FILTER_DATA_MGMT:
			filter_check = (frm_ctrl->type == NRF_WIFI_DATA_PKT_TYPE) ||
						   (frm_ctrl->type == NRF_WIFI_MGMT_PKT_TYPE);
			break;
		case FILTER_CTRL_ONLY:
			filter_check = (frm_ctrl->type == NRF_WIFI_CTRL_PKT_TYPE);
			break;
		case FILTER_CTRL_MGMT:
			filter_check = (frm_ctrl->type == NRF_WIFI_CTRL_PKT_TYPE) ||
						   (frm_ctrl->type == NRF_WIFI_MGMT_PKT_TYPE);
			break;
		case FILTER_DATA_CTRL:
			filter_check = (frm_ctrl->type == NRF_WIFI_DATA_PKT_TYPE) ||
						   (frm_ctrl->type == NRF_WIFI_CTRL_PKT_TYPE);
			break;
		default:
			// Allow all packet types if bit 0 is set or if all packet type bits are enabled
			filter_check = true;
			break;
	}
	return filter_check;
}
```

Reply:
yes ...believe it or not, I had this change made. I however had some issue cropping up wherein the filtering was going awry. Since I had a working solution, I decided to go with the safer approach. It was a conscious decision to keep it the way it is for the moment.

Thanks and regards
Vivek

Reply:
I understand, but that code is tough to read esp. with all those magic numbers.

Reply:
I provided detailed comments but I guess I will move them to an enum. The `if checks` , I will probably keep for the moment. 

Thanks and regards
Vivek